File: jconsamp/CGI-BIN/appsguid/AppsGuid.c
/*
(c) Copyright IBM Corp. 1997  All rights reserved.

These sample programs are owned by International Business Machines
Corporation or one of its subsidiaries ("IBM") and are copyrighted and
licensed, not sold.

You may copy, modify, and distribute these sample programs in any
form without payment to IBM,  for any purpose including developing,
using, marketing or distributing programs that include or are
derivative works of the sample programs.

The sample programs are provided to you on an "AS IS" basis, without
warranty of any kind.  IBM HEREBY EXPRESSLY DISCLAIMS ALL WARRANTIES,
EITHER EXPRESS OR IMPLIED, INCLUDING , BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
Some jurisdictions do not allow for the exclusion or limitation of
implied warranties, so the above limitations or exclusions may not
apply to you.  IBM shall not be liable for any damages you suffer
as a result of using, modifying or distributing the sample programs
or their derivatives.

Each copy of any portion of this/these sample program(s) or any
derivative work, must include a the above copyright notice and warranty.
*/



#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys\timeb.h>
#include "appsguid.h"

char **getcgivars() ;

int main ( int argc, char* argv[], char* envp[] )
{
	int i ;
	char **cgivars ;
	char szEvent[9] ;
	char szListItem [53] ;
	char szButton[9] ;

	
	strcpy( szListItem, "unknown" ) ;
	strcpy( szButton, "unknown" ) ;
	strcpy( szEvent, argv[1] ) ;

	if( !strcmp( szEvent, "init" ) )
	{
		BOOL sReturnCode ;
		char * pBuffer ;
		long int count ;
		char * pszToken ;
		struct timeb tstart, tstop ;
		unsigned short tstartmsecs, tstopmsecs ;
		unsigned short tused ;


		sReturnCode = connect( "appsguid", "mnix", "password" ) ;
		ftime( &tstart ) ;
		count = queryCount( ) ;	
		pBuffer = (char *) malloc( count * MAX_ID_AND_NAME_LENGTH ) ;
		memset(pBuffer, 0, count * MAX_ID_AND_NAME_LENGTH ) ;
		sReturnCode = queryProductIDsAndNames( &count, pBuffer ) ;
		ftime( &tstop ) ;
		tstartmsecs = ( 1000 * tstart.time ) + tstart.millitm ;
		tstopmsecs = ( 1000 * tstop.time ) + tstop.millitm ;
		tused =  tstopmsecs - tstartmsecs ;	//: returned in seconds

		
		printf( "Content-Type:  text/html\n\n" ) ;
		printf( "<HTML><HEAD><TITLE>" ) ;
		printf( "Solution Developer Marketing Application Guide" ) ;
		printf( "</TITLE></HEAD>" ) ;
		printf( "<BODY>" ) ;
		printf( "<H3>Applications Available for Free Download</H3>" ) ;
		printf( "<ul>" ) ;
		printf( "<FORM METHOD=\"POST\" ACTION=\"/cgi-bin/appsguid.exe?action\">" ) ;
		printf( "<TABLE COLS=2 FRAME=ALL WIDTH=600 NOWRAP>" ) ;
		printf( "<TBODY>" ) ;
		printf( "<TR>Select a product and press REFRESH button:" ) ;
		printf( "<TD NOWRAP>" ) ;
		printf( "<SELECT NAME=\"List\" SIZE=\"8\" MAXLENGTH=\"52\">" ) ;
		pszToken = strtok( pBuffer, ";" ) ;
		while( pszToken != NULL ) 
		{
			printf( "<OPTION>%s", pszToken ) ;
			pszToken = strtok(NULL, ";" ) ;
		}
		printf( "</SELECT>" ) ;
		printf( "<TD>" ) ;
		printf( "<TABLE COLS=2 FRAME=ALL WIDTH=600 NOWRAP>" ) ;
		printf( "<TR>" ) ;
		printf( "<TD><img src=\"http://dbserver/blueball.gif\">Product Name:" ) ;
		printf( "<TD>" ) ;

		printf( "<TR>" ) ;
		printf( "<TD><img src=\"http://dbserver/blueball.gif\">Product ID:" ) ;
		printf( "<TD>" ) ;

		printf( "<TR>" ) ;
		printf( "<TD><img src=\"http://dbserver/blueball.gif\">Description:" ) ;
		printf( "<TD>" ) ;

		printf( "<TR>" ) ;
		printf( "<TD><img src=\"http://dbserver/blueball.gif\">Supplier ID:" ) ;
		printf( "<TD>" ) ;
		printf( "</TABLE>" ) ;

		printf( "</TABLE>" ) ;
		printf( "<BR><INPUT NAME=\"Button\" TYPE=\"SUBMIT\" VALUE=\"Refresh\">" ) ;
		printf( "<SP><SP><INPUT NAME=\"Button\" TYPE=\"SUBMIT\" VALUE=\"Order\">" ) ;
		printf( "</FORM>" ) ;
		printf( "<BR>" ) ;
		printf( "<a href=\"http://www.ibm.com\">Press here to goto product's home page.</a>" ) ;
		printf( "</ul>" ) ;
		printf( "<BR><BR>Note: it took %ld milliseconds to load data.", tused ) ;
		printf( "</BODY></HTML>" ) ;

	}
	else //: Must be an action
	{
		i = 0 ;
           	//: Get the CGI variables
    		cgivars= getcgivars() ;

	


		
		//: Determine passed-in parameters 
    		for ( i = 0; cgivars[i]; i += 2)
		{
			if( !strcmp( cgivars[i], "Button" ) )
			{
				strcpy( szButton, cgivars[i+1] ) ;
			}

			if( !strcmp( cgivars[i], "List" ) )
			{
				strcpy( szListItem, cgivars[i+1] ) ;
			} 
		}
	
		

		//: Take action
		if( !strcmp( szButton, "Refresh" ) )
		{
			BOOL sReturnCode ;
			char * pBuffer ;
			char szProductID[5] ;
			PProduct pProduct = (PProduct) malloc( sizeof( Product ) ) ;
			long int count ;
			char * pszToken ;
			struct timeb tstart, tstop ;
			unsigned short tstartmsecs, tstopmsecs ;
			unsigned short tused ;


			if( !strcmp( szListItem, "unknown" ) )
				exit( 1 ) ;
			memset( pProduct, 0, sizeof( Product ) ) ;
			strncpy( pProduct->product_id, szListItem, 4 ) ;
	

			sReturnCode = connect( "appsguid", "mnix", "password" ) ;
			sReturnCode = getProduct( pProduct ) ;
			ftime( &tstart ) ;
			count = queryCount( ) ;	
			pBuffer = (char *) malloc( count * MAX_ID_AND_NAME_LENGTH ) ;
			memset(pBuffer, 0, count * MAX_ID_AND_NAME_LENGTH ) ;
			sReturnCode = queryProductIDsAndNames( &count, pBuffer ) ;
			ftime( &tstop ) ;
			tstartmsecs = ( 1000 * tstart.time ) + tstart.millitm ;
			tstopmsecs = ( 1000 * tstop.time ) + tstop.millitm ;
			tused =  tstopmsecs - tstartmsecs ;	//: returned in seconds


			printf( "Content-Type:  text/html\n\n" ) ;
			printf( "<HTML><HEAD><TITLE>" ) ;
			printf( "Solution Developer Marketing Application Guide" ) ;
			printf( "</TITLE></HEAD>" ) ;
			printf( "<BODY>" ) ;
			printf( "<H3>Applications Available for Free Download</H3>" ) ;
			printf( "<ul>" ) ;
			printf( "<FORM METHOD=\"POST\" ACTION=\"/cgi-bin/appsguid.exe?action\">" ) ;
			printf( "<TABLE COLS=2 FRAME=ALL WIDTH=600 NOWRAP>" ) ;
			printf( "<TBODY>" ) ;
			printf( "<TR>Select a product and press REFRESH button:" ) ;
			printf( "<TD NOWRAP>" ) ;
			printf( "<SELECT NAME=\"List\" SIZE=\"8\" MAXLENGTH=\"52\">" ) ;
			pszToken = strtok( pBuffer, ";" ) ;
			while( pszToken != NULL ) 
			{
				printf( "<OPTION>%s", pszToken ) ;
				pszToken = strtok(NULL, ";" ) ;
			}
			printf( "</SELECT>" ) ;
			printf( "<TD>" ) ;
			printf( "<TABLE COLS=2 FRAME=ALL WIDTH=600 NOWRAP>" ) ;
			printf( "<TR>" ) ;
			printf( "<TD><img src=\"http://dbserver/blueball.gif\">Product Name:" ) ;
			printf( "<TD>%s", pProduct->product_name  ) ;
	
			printf( "<TR>" ) ;
			printf( "<TD><img src=\"http://dbserver/blueball.gif\">Product ID:" ) ;
			printf( "<TD>%s", pProduct->product_id ) ;
	
			printf( "<TR>" ) ;
			printf( "<TD><img src=\"http://dbserver/blueball.gif\">Description:" ) ;
			printf( "<TD>%s", pProduct->description ) ;

			printf( "<TR>" ) ;
			printf( "<TD><img src=\"http://dbserver/blueball.gif\">Supplier ID:" ) ;
			printf( "<TD>%s", pProduct->supplier_id ) ;
			printf( "</TABLE>" ) ;

			printf( "</TABLE>" ) ;
			printf( "<BR><INPUT NAME=\"Button\" TYPE=\"SUBMIT\" VALUE=\"Refresh\">" ) ;
			printf( "<SP><SP><INPUT NAME=\"Button\" TYPE=\"SUBMIT\" VALUE=\"Order\">" ) ;
			printf( "</FORM>" ) ;
			printf( "<BR>" ) ;
			printf( "<a href=\"%s\">Press here to goto product's home page.</a>", pProduct->url ) ;
			printf( "</ul>" ) ;
			printf( "<BR><BR>Note: it took %ld milliseconds to load data.", tused ) ;
			printf( "</BODY></HTML>" ) ;

	
		}

		else if( !strcmp( szButton, "Order" ) ) 
		{
		}

		else
		{
			printf( "Content-Type:  text/html\n\n" ) ;
			printf( "<HTML><HEAD><TITLE>" ) ;
			printf( "Solution Developer Marketing Application Guide" ) ;
			printf( "</TITLE></HEAD>" ) ;
			printf( "<BODY>" ) ;
			printf( "Missed the boat" ) ;
			printf( "</BODY></HTML>" ) ;

		}

     
      
	     	
	}

	exit( 0 ) ;
}


/** Read the CGI input and place all name/val pairs into list.        **/
/** Returns list containing name1, value1, name2, value2, ... , NULL  **/
char **getcgivars() {
    register int i ;
    char *request_method ;
    int content_length;
    char *cgiinput ;
    char **cgivars ;
    char **pairlist ;
    int paircount ;
    char *nvpair ;
    char *eqpos ;

    /** Depending on the request method, read all CGI input into cgiinput **/
    /** (really should produce HTML error messages, instead of exit()ing) **/
    request_method= getenv("REQUEST_METHOD") ;
    if (!strcmp(request_method, "GET")) {
        cgiinput= strdup(getenv("QUERY_STRING")) ;
    }
    else if (!strcmp(request_method, "POST")) {
        /* strcasecmp() is not supported in Windows-- use strcmpi() instead */
        if ( strcmpi(getenv("CONTENT_TYPE"), "application/x-www-form-urlencoded")) {
            printf("getcgivars(): Unsupported Content-Type.\n") ;
            exit(1) ;
        }
        if ( !(content_length = atoi(getenv("CONTENT_LENGTH"))) ) {
            printf("getcgivars(): No Content-Length was sent with the POST request.\n") ;
            exit(1) ;
        }
        if ( !(cgiinput= (char *) malloc(content_length+1)) ) {
            printf("getcgivars(): Could not malloc for cgiinput.\n") ;
            exit(1) ;
        }
        if (!fread(cgiinput, content_length, 1, stdin)) {
            printf("Couldn't read CGI input from STDIN.\n") ;
            exit(1) ;
        }
        cgiinput[content_length]='\0' ;
    }
    else {
        printf("getcgivars(): unsupported REQUEST_METHOD\n") ;
        exit(1) ;
    }

    /** Change all plusses back to spaces **/
    for(i=0; cgiinput[i]; i++) if(cgiinput[i] == '+') cgiinput[i] = ' ' ;

    /** First, split on "&" to extract the name-value pairs into pairlist **/
    pairlist= (char **) malloc(256*sizeof(char **)) ;
    paircount= 0 ;
    nvpair= strtok(cgiinput, "&") ;
    while (nvpair) {
        pairlist[paircount++]= strdup(nvpair) ;
        if (!(paircount%256))
            pairlist= (char **) realloc(pairlist,(paircount+256)*sizeof(char **)) ;
        nvpair= strtok(NULL, "&") ;
    }
    pairlist[paircount]= 0 ;    /* terminate the list with NULL */

    /** Then, from the list of pairs, extract the names and values **/
    cgivars= (char **) malloc((paircount*2+1)*sizeof(char **)) ;
    for (i= 0; i<paircount; i++) {
        if (eqpos=strchr(pairlist[i], '=')) {
            *eqpos= '\0' ;
            cgivars[i*2+1]= strdup(eqpos+1) ;
        } else {
            cgivars[i*2+1]= strdup("") ;
        }
        cgivars[i*2]= strdup(pairlist[i]) ;
    }
    cgivars[paircount*2]= 0 ;   /* terminate the list with NULL */
    
    /** Free anything that needs to be freed **/
    free(cgiinput) ;
    for (i=0; pairlist[i]; i++) free(pairlist[i]) ;
    free(pairlist) ;

    /** Return the list of name-value strings **/
    return cgivars ;
    
}

File: jconsamp/Servlets/AppsGuid/AppsGuid.java
/*
(c) Copyright IBM Corp. 1997  All rights reserved.

These sample programs are owned by International Business Machines
Corporation or one of its subsidiaries ("IBM") and are copyrighted and
licensed, not sold.

You may copy, modify, and distribute these sample programs in any
form without payment to IBM,  for any purpose including developing,
using, marketing or distributing programs that include or are
derivative works of the sample programs.

The sample programs are provided to you on an "AS IS" basis, without
warranty of any kind.  IBM HEREBY EXPRESSLY DISCLAIMS ALL WARRANTIES,
EITHER EXPRESS OR IMPLIED, INCLUDING , BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
Some jurisdictions do not allow for the exclusion or limitation of
implied warranties, so the above limitations or exclusions may not
apply to you.  IBM shall not be liable for any damages you suffer
as a result of using, modifying or distributing the sample programs
or their derivatives.

Each copy of any portion of this/these sample program(s) or any
derivative work, must include a the above copyright notice and warranty.
*/



package com.ibm.sdm.appsguid ;

import java.io.IOException ;
import java.io.PrintStream ;
import java.io.PrintWriter ;

import java.util.Enumeration ;
import java.util.Hashtable ;

import javax.servlet.ServletConfig ;
import javax.servlet.ServletContext ;
import javax.servlet.ServletRequest ;
import javax.servlet.ServletResponse ;
import javax.servlet.ServletInputStream ;
import javax.servlet.ServletOutputStream ;
import javax.servlet.ServletException ;

import javax.servlet.http.HttpServlet ;
import javax.servlet.http.HttpServletRequest ;
import javax.servlet.http.HttpServletResponse ;
import javax.servlet.http.HttpUtils ;
import javax.servlet.http.HttpSession ;

import com.ibm.ServletExpress.personalization.sessiontracking.IBMSessionData;
// import com.ibm.ServletExpress.personalization.userprofile.*;
import com.ibm.ServletExpress.personalization.util.VariableSubstitution ;
// import com.ibm.ServletExpress.util.*  ;
//import com.sun.server.http.session.*;


import java.sql.*;   
import COM.ibm.db2.jdbc.app.*;
import COM.ibm.db2.app.*;   



public class AppsGuid extends javax.servlet.http.HttpServlet
{
	private IBMSessionData session ;
	private Enumeration initParms ;
	private ServletContext context ;	
	private static ConnectionManager connMgr ;
	private Hashtable hashtable ;
	private StringBuffer buffer ;
	//: Place html file in Web server's public html root...
	public static final String htmlUrl  = "ProductInformation.html";

	//: Initializer section
	static
	{
		 try
        	  {
	   		Class.forName ("COM.ibm.db2.jdbc.app.DB2Driver").newInstance () ;
           		connMgr = new ConnectionManager();
           		System.out.println("Connections pool created");
        	  } //end  try
        	catch (Exception exception)
        	  {
           		exception.printStackTrace();
        	  } // end catch
        

	}


	//:  JavaSoft Java Servlet API 1.0 does not allow an init method with no function arguments
	public void init( ServletConfig config ) 
	{
		try {	
			super.init( config ) ; //:  Must call if overriding init( ServletConfig ) ;
			hashtable = new Hashtable() ;
			context = getServletContext( ) ;
			Connection connection = null ;
			try
		  	{
				connection = connMgr.getConnection();
				String sql = "Call GET_AM_ID_AND_NAME(?) " ;
				CallableStatement statement = connection.prepareCall(sql) ;
				statement.registerOutParameter (1, Types.CHAR);  //: Other
				String container = "sample" ;
			
				statement.setString(1,container) ;		//: setObject
				statement.execute() ;
			
				container = statement.getString(1) ;

				//: Parse info

				String temp = container ;
			
				int index = temp.indexOf(':') ;
				String numberString = temp.substring(0, index) ;
				int numberOfItems = Integer.parseInt(numberString) ;
				temp = temp.substring(index + 1) ;
				for( int i=0; i < numberOfItems; i++)
				{
					index = temp.indexOf(':') ;
					String product_id = temp.substring(0,index) ;
					
					temp = temp.substring(index + 1) ;
					index = temp.indexOf(';') ;
					String product_name = temp.substring(0, index) ;
					hashtable.put(product_id, product_name) ;
					temp = temp.substring(index + 1) ;
				}	

			
			
		  	}
			catch(Exception exception)
			  {
				exception.printStackTrace() ;
			  }
			finally
			  {
				try
				  {
					connMgr.putConnection(connection);
			  	} catch(Exception excpt) { excpt.printStackTrace() ; }
		  	  }


			initParms = config.getInitParameterNames( ) ;
			//:  ServletContext::log writes to X:\ServletExpress\logs\ServletExpress\servletservice\event_log
			context.log( "AppsGuid init( ServletConfig ) method invoked.\n" ) ;
		} catch( Exception exception ) {
			context.log( exception, exception.toString( ) ) ;
		}
	}
	


	public void destroy()
	{
		try
        	{
          		connMgr.destroyPool();
			connMgr = null ;
        	}
        	catch (Exception exception)
        	{
           		exception.printStackTrace();
			connMgr = null ;
        	} // end catch
	}
/*
  
	//:  Could override service( ServletRequest, ServletResponse ) ;but, 
	//:	you'll lose default TRACE and OUTPUT methods processing.
	//:	Notice that the service method is not typically overridden. The service method, as provided,
	//:	supports standard HTTP requests by dispatching them to appropriate methods, such as the methods
	//:	listed above that have the prefix "do". In addition, the service method also supports the HTTP 1.1
	//:	protocol's TRACE and OPTIONS methods by dispatching to the doTrace and doOptions
	//:	methods. The doTrace and doOptions methods are not typically overridden. 
	public void service( ServletRequest request, ServletResponse response ) throws ServletException, IOException
	{
		try {
			context.log("Init Parms are:  ") ;
			while( initParms.hasMoreElements( ) )
			{
				context.log((String)initParms.nextElement( ) ) ;
			}
			context.log( context.getServerInfo( ) ) ;
			Enumeration enum = context.getServletNames( ) ;
			while( enum.hasMoreElements( ) )
			{
				context.log( (String)enum.nextElement( ) ) ;
			}

		} catch( Exception exception ) {
			context.log( exception, (String)exception.toString( ) ) ;
		}

		super.service(request, response) ;
	}

  
*/

	//:  Performs the HTTP GET operation; the default implementation 
	//:	reports an HTTP BAD_REQUEST error.
	protected void doGet(HttpServletRequest request,
                     HttpServletResponse response) throws ServletException, IOException 
	{
		Connection connection = null ;

		session = (IBMSessionData) request.getSession(true);
		context.log( "Using improvements...") ;
		context.log( "Got the HTTP_GET request.") ;
		//:  Invoke servlet with http://host/servlet/ServletTemplate?FN=Mickey&LN=Nix
		context.log(request.getQueryString( ) ) ;
	
	
		//: Add some real HTML
		//: Probably don't want to use "text/plain" with HTML tags
		//: WARNING:  DON'T TRY TO MIX "text/XXX" types in same servlet,
		//:	as you'll get a 404 error
		response.setContentType( "text/html" ) ;

		PrintWriter print = response.getWriter( ) ;
		
		print.println( "<HTML><HEAD><TITLE>" ) ;
		print.println( "Solution Developer Marketing Application Guide" ) ;
		print.println( "</TITLE></HEAD>" ) ;
		print.println( "<BODY>" ) ;
		print.println( "<H3>Applications Available for Free Download</H3>" ) ;
		print.println( "<ul>" ) ;
		print.println( "<FORM METHOD=\"POST\" ACTION=\"/servlet/AppsGuid\">" ) ;
		
		print.println( "<BR>Select a product and press REFRESH button:" ) ;
		print.println( "<BR><BR><SELECT NAME=\"List\" SIZE=\"8\" MAXLENGTH=\"52\">" ) ;
		
		if(hashtable.isEmpty() == false)
		  {
			for(Enumeration enum = hashtable.keys() ; enum.hasMoreElements() ; )
			  {
				String key = (String)enum.nextElement() ;
				String value = (String)hashtable.get(key) ;
				print.println( "<OPTION>" + key + " : " + value ) ;
			  }
		  }
		print.println( "</SELECT>" ) ;
		
		print.println( "<BR><BR><INPUT NAME=\"Button\" TYPE=\"SUBMIT\" VALUE=\"Refresh\">" ) ;
		print.println( "<SP><SP><INPUT NAME=\"Button\" TYPE=\"SUBMIT\" VALUE=\"Order\">" ) ;
		print.println( "</FORM>" ) ;
		
		print.println( "</ul>" ) ;
	
		print.println( "</BODY></HTML>" ) ;

		print.flush( ) ;
		print.close( ) ;
	
	}




	//:  Performs the HTTP POST operation; the default implementation 
	//:	reports an HTTP BAD_REQUEST error.
	protected void doPost(HttpServletRequest request,
                     HttpServletResponse response) throws ServletException, IOException 
	{
		Connection connection ;

		context.log( "Got the HTTP_POST request.") ;
		context.log("Post query string is " + request.getQueryString( ) ) ;
	
		//: Get the current sessionid - starts a session, if one does not exist yet.
       		session = (IBMSessionData) request.getSession(true);
		response.setContentType( "text/html" ) ;
	
		PrintWriter print = response.getWriter( ) ;
			

		Enumeration enum = request.getParameterNames( ) ;
		while(enum.hasMoreElements( ) ) 
		{
			String name = (String)enum.nextElement() ;
			String value[] = request.getParameterValues(name);
			for(int i=0; i < value.length; i++)
			{
				if(name.equals("Button") && value[i].equals("Refresh"))
				{
					try
					  {
						String list[] = request.getParameterValues("List") ;
						String product_id = list[0].substring(0,4) ;
						String product_name = list[0].substring(7) ;
					
						connection = connMgr.getConnection();
						String sql = "Call GET_AM_PRODUCT(?,?) " ;
						CallableStatement statement = connection.prepareCall(sql) ;

						statement.registerOutParameter(2, Types.CHAR) ;
						
						statement.setString(1,product_id) ;
						String unknown = new String("unkown") ;
						statement.setString(2,unknown) ;
						statement.execute() ;
						
						String info = statement.getString(2) ;
						
						int index = info.indexOf('\t') ;
						info = info.substring(index + 1) ;
						index = info.indexOf('\t') ;
						product_name = info.substring(0, index) ; 
						session.putValue(new String("product_id"), product_id);
						session.putValue(new String("product_name"), product_name);
											
						info = info.substring(index + 1) ;
						index = info.indexOf('\t') ;
						String description = info.substring(0, index) ;
						session.putValue(new String("description"), description);
						
						info = info.substring(index + 1) ;
						index = info.indexOf('\t') ;
						String supplier_id = info.substring(0, index) ;
						session.putValue(new String("supplier_id"), supplier_id);

						String url = info.substring(index + 1) ;
						session.putValue(new String("url"), url) ;


						
						statement.close() ;
						//: Variable substitution--see <!META VAR...> tags above

						StringBuffer output = VariableSubstitution.substitute(htmlUrl, null, request);
						print.println(new String(output));
     						print.flush();
     						print.close();

					  } catch(Exception exception) { exception.printStackTrace() ; }
				}
			}
		}
	

	
		print.println( "</BODY></HTML>" ) ;
		print.flush() ;
		print.close() ;

          
	}


  
}
